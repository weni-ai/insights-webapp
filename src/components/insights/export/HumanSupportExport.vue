<template>
  <section class="modal-export-data">
    <UnnnicToolTip
      data-testid="export-data-tooltip"
      :text="t('export_data.tooltip')"
      side="left"
      :enabled="!hasExportData"
    >
      <UnnnicButton
        type="primary"
        size="large"
        :text="t('export_data.title')"
        :loading="isLoadingCheckExportStatus"
        :disabled="!hasExportData"
        @click="setIsRenderExportData(true)"
      />
    </UnnnicToolTip>
    <UnnnicModalDialog
      data-test-id="modal-dialog"
      :modelValue="isRenderExportData"
      class="finish-onboarding-modal"
      :primaryButtonProps="{
        text: t('export_data.save_btn'),
        disabled: !hasEnabledToExport,
        loading: isLoadingCreateExport,
      }"
      :secondaryButtonProps="{ text: t('export_data.cancel_btn') }"
      showActionsDivider
      size="lg"
      :title="t('export_data.title')"
      showCloseIcon
      @primary-button-click="createExport"
      @secondary-button-click="setIsRenderExportData(false)"
      @update:model-value="setIsRenderExportData(false)"
    >
      <FormExport />
    </UnnnicModalDialog>
    <UnnnicModalDialog
      data-test-id="modal-dialog-feedback"
      :modelValue="isRenderExportDataFeedback"
      class="finish-onboarding-modal"
      :title="t('export_data.feedback_title')"
      type="success"
      size="md"
      showCloseIcon
      @update:model-value="setIsRenderExportDataFeedback(false)"
    >
      <p class="export-data-feedback__text">
        {{ $t('export_data.feedback') }}
      </p>
    </UnnnicModalDialog>
  </section>
</template>

<script setup lang="ts">
import { storeToRefs } from 'pinia';
import { useI18n } from 'vue-i18n';
import { onMounted, onUnmounted, ref, computed } from 'vue';
import { useHumanSupportExport } from '@/store/modules/export/humanSupport/export';
import FormExport from './HumanSupport/FormExport.vue';

const { t } = useI18n();
const humanSupportExport = useHumanSupportExport();
const { setIsRenderExportData, setIsRenderExportDataFeedback, createExport } =
  humanSupportExport;
const {
  isRenderExportData,
  isRenderExportDataFeedback,
  hasEnabledToExport,
  isLoadingCreateExport,
  export_data,
  isLoadingCheckExportStatus,
} = storeToRefs(humanSupportExport);

const pollingInterval = ref<ReturnType<typeof setInterval> | null>(null);
const secondsToPoll = ref(60000);

const hasExportData = computed(() => {
  return export_data.value?.status?.toLowerCase() === 'ready';
});

const startPolling = () => {
  if (pollingInterval.value) {
    clearInterval(pollingInterval.value);
  }

  pollingInterval.value = setInterval(() => {
    humanSupportExport.checkExportStatus();
  }, secondsToPoll.value);
};

const stopPolling = () => {
  if (pollingInterval.value) {
    clearInterval(pollingInterval.value);
    pollingInterval.value = null;
  }
};

onMounted(() => {
  humanSupportExport.checkExportStatus();
  startPolling();
});

onUnmounted(() => {
  stopPolling();
});
</script>

<style lang="scss" scoped>
.modal-export-data {
  display: flex;
}

.export-data-feedback__text {
  font-family: $unnnic-font-family-secondary;
  color: $unnnic-color-neutral-cloudy;
  font-size: $unnnic-font-size-body-gt;
  font-weight: $unnnic-font-weight-regular;
  line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
}
</style>
