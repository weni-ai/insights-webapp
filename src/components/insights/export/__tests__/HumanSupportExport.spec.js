import { beforeEach, afterEach, describe, expect, it, vi } from 'vitest';
import { config, mount } from '@vue/test-utils';
import { createI18n } from 'vue-i18n';

import HumanSupportExport from '../HumanSupportExport.vue';

const mockStore = {
  isRenderExportData: { value: false },
  isRenderExportDataFeedback: { value: false },
  hasEnabledToExport: { value: true },
  isLoadingCreateExport: { value: false },
  isLoadingCheckExportStatus: { value: false },
  export_data: { value: { status: 'ready' } },
  setIsRenderExportData: vi.fn(),
  setIsRenderExportDataFeedback: vi.fn(),
  createExport: vi.fn(),
  checkExportStatus: vi.fn(),
};

vi.mock('@/store/modules/export/humanSupport/export', () => ({
  useHumanSupportExport: () => mockStore,
}));

vi.mock('pinia', async (importOriginal) => {
  const actual = await importOriginal();
  return {
    ...actual,
    storeToRefs: (store) => store,
  };
});

const i18n = createI18n({
  legacy: false,
  locale: 'en',
  messages: {
    en: {
      export_data: {
        title: 'Export Data',
        tooltip: 'Tooltip text',
        save_btn: 'Save',
        cancel_btn: 'Cancel',
        feedback_title: 'Success',
        feedback: 'Export successful',
      },
    },
  },
});

config.global.plugins = [i18n];

describe('HumanSupportExport', () => {
  let wrapper;

  const createWrapper = (storeOverrides = {}) => {
    Object.assign(mockStore, storeOverrides);

    return mount(HumanSupportExport, {
      global: {
        stubs: {
          UnnnicButton: true,
          UnnnicToolTip: true,
          UnnnicModalDialog: true,
          FormExport: true,
        },
      },
    });
  };

  const section = () => wrapper.find('[data-testid="modal-export-data"]');

  beforeEach(() => {
    vi.useFakeTimers();
    vi.clearAllMocks();

    Object.assign(mockStore, {
      isRenderExportData: { value: false },
      isRenderExportDataFeedback: { value: false },
      hasEnabledToExport: { value: true },
      isLoadingCreateExport: { value: false },
      isLoadingCheckExportStatus: { value: false },
      export_data: { value: { status: 'ready' } },
    });

    wrapper = createWrapper();
  });

  afterEach(() => {
    vi.useRealTimers();
    wrapper.unmount();
  });

  describe('Component rendering', () => {
    it('should render main section', () => {
      expect(section().exists()).toBe(true);
    });

    it('should render tooltip', () => {
      const tooltip = wrapper.find('[data-testid="export-data-tooltip"]');
      expect(tooltip.exists()).toBe(true);
    });
  });

  describe('Computed properties', () => {
    it('should compute hasExportData as true when status is ready', () => {
      wrapper = createWrapper({
        export_data: { value: { status: 'ready' } },
      });

      expect(wrapper.vm.hasExportData).toBe(true);
    });

    it('should compute hasExportData as true when status is READY (uppercase)', () => {
      wrapper = createWrapper({
        export_data: { value: { status: 'READY' } },
      });

      expect(wrapper.vm.hasExportData).toBe(true);
    });

    it('should compute hasExportData as false when status is not ready', () => {
      wrapper = createWrapper({
        export_data: { value: { status: 'processing' } },
      });

      expect(wrapper.vm.hasExportData).toBe(false);
    });

    it('should handle null export_data', () => {
      wrapper = createWrapper({
        export_data: { value: null },
      });

      expect(wrapper.vm.hasExportData).toBe(false);
    });
  });

  describe('Lifecycle hooks', () => {
    it('should call checkExportStatus on mounted', () => {
      expect(mockStore.checkExportStatus).toHaveBeenCalled();
    });

    it('should start polling on mounted', () => {
      vi.advanceTimersByTime(60000);
      expect(mockStore.checkExportStatus).toHaveBeenCalledTimes(2);
    });

    it('should stop polling on unmounted', () => {
      const checkCallCount = mockStore.checkExportStatus.mock.calls.length;
      wrapper.unmount();

      vi.advanceTimersByTime(60000);
      expect(mockStore.checkExportStatus).toHaveBeenCalledTimes(checkCallCount);
    });
  });

  describe('Polling logic', () => {
    it('should poll every 60 seconds', () => {
      const initialCalls = mockStore.checkExportStatus.mock.calls.length;

      vi.advanceTimersByTime(60000);
      expect(mockStore.checkExportStatus).toHaveBeenCalledTimes(
        initialCalls + 1,
      );

      vi.advanceTimersByTime(60000);
      expect(mockStore.checkExportStatus).toHaveBeenCalledTimes(
        initialCalls + 2,
      );
    });

    it('should clear previous interval when starting new polling', () => {
      wrapper.vm.startPolling();
      const firstInterval = wrapper.vm.pollingInterval;

      wrapper.vm.startPolling();
      const secondInterval = wrapper.vm.pollingInterval;

      expect(firstInterval).not.toBe(secondInterval);
    });

    it('should clear interval on stopPolling', () => {
      wrapper.vm.stopPolling();
      expect(wrapper.vm.pollingInterval).toBeNull();
    });
  });

  describe('Modal dialogs', () => {
    it('should render export modal', () => {
      const modal = wrapper.find('[data-testid="modal-dialog"]');
      expect(modal.exists()).toBe(true);
    });

    it('should render export modal with correct value when open', () => {
      wrapper = createWrapper({
        isRenderExportData: { value: true },
      });

      const modal = wrapper.find('[data-testid="modal-dialog"]');
      expect(modal.exists()).toBe(true);
    });

    it('should render feedback modal', () => {
      const modal = wrapper.find('[data-testid="modal-dialog-feedback"]');
      expect(modal.exists()).toBe(true);
    });

    it('should render feedback modal with correct value when open', () => {
      wrapper = createWrapper({
        isRenderExportDataFeedback: { value: true },
      });

      const modal = wrapper.find('[data-testid="modal-dialog-feedback"]');
      expect(modal.exists()).toBe(true);
    });
  });

  describe('Component structure', () => {
    it('should have correct CSS classes', () => {
      expect(section().classes()).toContain('modal-export-data');
    });

    it('should match snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
