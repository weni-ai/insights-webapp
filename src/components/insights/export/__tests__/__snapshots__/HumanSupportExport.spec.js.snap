// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HumanSupportExport > Component structure > should match snapshot 1`] = `
<section
  class="modal-export-data"
  data-testid="modal-export-data"
  data-v-81b4e24b=""
>
  <unnnic-tool-tip-stub
    data-testid="export-data-tooltip"
    data-v-81b4e24b=""
    enabled="false"
    side="left"
    text="Tooltip text"
  />
  <unnnic-modal-dialog-stub
    class="finish-onboarding-modal"
    data-testid="modal-dialog"
    data-v-81b4e24b=""
    modelvalue="[object Object]"
    primarybuttonprops="[object Object]"
    secondarybuttonprops="[object Object]"
    showactionsdivider=""
    showcloseicon=""
    size="lg"
    title="Export Data"
  />
  <unnnic-modal-dialog-stub
    class="finish-onboarding-modal"
    data-testid="modal-dialog-feedback"
    data-v-81b4e24b=""
    modelvalue="[object Object]"
    showcloseicon=""
    size="md"
    title="Success"
    type="success"
  />
</section>
`;
